tiny_dnn/core/kernels/fully_connected_op_avx.h

```diff
@@ -14,225 +14,304 @@ namespace kernels {
+inline void avx_fully_connected_back_kernel(const Tensor<float, S1> &prev_out,
+                                            const Tensor<float, S2> &W,
+                                            Tensor<float, S3> &dW,
+                                            Tensor<float, S4> &db,
+                                            Tensor<float, S5> &curr_delta,
+                                            Tensor<float, S6> &prev_delta,
+                                            const bool has_bias,
+                                            const bool layer_parallelize) {
+  size_t out_size = curr_delta.shape()[1], in_size = prev_delta.shape()[1];
+  if (has_bias) {
+    for (serial_size_t sample = 0; sample < prev_out.shape()[0]; sample++) {
+      for (serial_size_t c = 0; c < in_size; c++) {
         // propagate delta to previous layer
         // prev_delta[c] += current_delta[r] * W_[c * out_size_ + r]
-        prev_delta2[c] += vectorize::dot(
-          &curr_delta2[0], &W[c * params.out_size_], params.out_size_);
+        prev_delta.host_at(sample, c) +=
+          vectorize::dot(curr_delta.host_pointer(sample, 0),
+                         W.host_pointer(0, c * out_size), out_size);
       }
-      for_(layer_parallelize, 0, size_t(params.out_size_),
-           [&](const blocked_range &r) {
-             // accumulate weight-step using delta
-             // dW[c * out_size + i] += current_delta[i] * prev_out[c]
-             size_t len = r.end() - r.begin();
-             for (serial_size_t c = 0; c < params.in_size_; c++) {
-               vectorize::muladd(&curr_delta2[r.begin()], prev_out2[c], len,
-                                 &dW2[c * params.out_size_ + r.begin()]);
-             }
-             // vec_t& db = *in_grad[2];
-             vectorize::reduce(&curr_delta2[r.begin()], len, &db2[r.begin()]);
-           });
+      for_(layer_parallelize, 0, size_t(out_size), [&](const blocked_range &r) {
+        // accumulate weight-step using delta
+        // dW[c * out_size + i] += current_delta[i] * prev_out[c]
+        size_t len = r.end() - r.begin();
+        for (serial_size_t c = 0; c < in_size; c++) {
+          vectorize::muladd(curr_delta.host_pointer(sample, r.begin()),
+                            prev_out.host_at(sample, c), len,
+                            dW.host_pointer(sample, c * out_size + r.begin()));
+        }
+        // vec_t& db = *in_grad[2];
+        vectorize::reduce(curr_delta.host_pointer(sample, r.begin()), len,
+                          db.host_pointer(sample, r.begin()));
+      });
     }
   } else {
-    for (serial_size_t sample = 0; sample < prev_out.size(); sample++) {
-      auto &prev_delta2 = prev_delta[sample];
-      auto &curr_delta2 = curr_delta[sample];
-      auto &prev_out2   = prev_out[sample];
-      auto &dW2         = dW[sample];
-      for (serial_size_t c = 0; c < params.in_size_; c++) {
+    for (serial_size_t sample = 0; sample < prev_out.shape()[0]; sample++) {
```

Comment:
You can use local variables inside this loop just like it used to be.
There's no point in repeating costly operations when it can be avoided.